specification/basic_concepts/basic_concepts.tex

```diff
@@ -1,33 +1,62 @@
+For each type of communication, a predefined set of data types is used,
+where each data type has a unique name.
+Additionally, every data type definition has a pair of major and minor version numbers,
 which enable data type definitions to evolve in arbitrary ways while ensuring a well-defined
 migration path if backward-incompatible changes are introduced.
-Some data structures are standard and defined by the protocol specification (of which only a small 
+Some data types are standard and defined by the protocol specification (of which only a small
 subset are required); others may be specific to a particular application or vendor.
 
-Since every message or service data type has its own unique data type ID,
-and each node in the network has its own unique node ID,
-a pair of data type ID and node ID can be used to support redundant nodes with identical
-functionality inside the same network.
-
-Message and service data structures are defined using the Data Structure Description Language
-(DSDL) (chapter \ref{sec:dsdl}).
-A DSDL description can be used to automatically generate the serialization/deserialization code
-for every defined data structure in a particular programming language.
+Message exchanges between nodes are grouped into \emph{subjects} by their semantic meaning.
+Message exchanges belonging to the same subject use same message data type down to the major version
+(minor versions are allowed to differ) and pertain to same function or process within the system.
+
+Request/response exchanges between nodes are grouped into \emph{services} by their semantic meaning,
+like messages are grouped into subjects.
+Requests and their corresponding responses that belong to the same service use same service data type down to
+the major version (minor versions are allowed to differ; as a consequence, the minor data type version number
+of a service response may differ from that of its corresponding request) and pertain to same function.
+
+Each message subject is identified by a unique natural number -- a \emph{subject ID};
+likewise, each service is identified by a unique \emph{service ID}.
+An umbrella term \emph{port ID} is used to refer either to a subject ID or to a service ID
+(port identifiers have no direct manifestation in the construction of the protocol,
+but they are convenient for discussion).
+
+Message and service data types\footnote{Data types are the core concept of UAVCAN.}
+are defined using the \emph{data structure description language} (DSDL) (chapter \ref{sec:dsdl}).
+A DSDL definition specifies the name, major version, minor version, the data structure,
+and an optional predefined port ID of the data type.
+Message data types always contain exactly one data structure, whereas
+service data types contain two data structures: one for request, and the other for response.
+
+If a port identifier is defined by a data type definition, such identifier is said to be a
+\emph{regulated port identifier}
```

Comment:
The name is per this discussion; my understanding was that we've all agreed with (or at least tolerated) the word "regulated": https://forum.uavcan.org/t/on-subject-service-id-ranges/195/4?u=pavel.kirienko

Reply:
No it's quite different.

How you present it in the text is if I create an application specific definition with a standard suject ID of 11, 11 would be a regulated port identifier. And if IDs allocated for the standard types do not have a type which sets them as their standard ID they're unregulated.

How I expected it to be is that regulated IDs are IDs which is globally vetted for conflicts, while unregulated is done per application.

Reply:
> How you present it in the text is if I create an application specific definition with a standard suject ID of 11, 11 would be a regulated port identifier.

Yes, this is the intended meaning of the text (and your definition would be non-compliant and rejected by parsers). The background is presented in the first post here https://forum.uavcan.org/t/on-subject-service-id-ranges/195, specifically:

> The above has an implication that vendors can no longer define private DTS that contain static Subject/Service IDs, becase if that were the case, definitions from different vendors might conflict with each other and there would be no way of fixing that without modifying the DTS, because, see, the IDs are unchangeable. Public DTS, on the other hand, can rely on static ID as long as they are all centralized in one place, where the lack of collisions can be determined statically. Such centralized maintenance can be achieved by either assigning ID ranges to vendors (which seems wasteful, as the ID spaces are limited), or by keeping all vendor-specific definitions (in their respective namespaces) in one central repository, such as https://github.com/UAVCAN/vendor-specific-dsdl.
>
> If the above reasoning is found sound, we are to end up with the following kinds of DTS:
>
> - Standard DTS, where identifiers can be static (conflicts are not possible because the set is the same for everyone) or dynamic.
> - Vendor-specific public DTS, where the identifiers can be static (conflicts are avoidable if we keep all definitions in one central repository or assign per-vendor ranges) or dynamic.
> - Vendor-specific private DTS, where the identifiers can be only dynamic, because vendors wouldn’t be able to guarantee collision-free coexistence of their data types.

To be more specific, you can't really define a data type with a regulated subject ID set to 11, because it lies outside the regulated ranges we discussed. PyDSDL checks for that specifically and throws an error.

This might seem restrictive, as you won't be able to define your own private application-specific data types with fixed identifiers, but this is necessary to statically guarantee collision-free coexistence of nodes from different vendors that use private data types. The core idea is that if you want a fixed ID, it must be public, so that other vendors can see it and avoid reusing it.

Reply:
I thought this sorted out by the "this range can also be used with static IDS post" later on. By rejecting definitions of default service/subject ID in the application specific domain will make the life of groups with full control of the nodes harder without gaining any security as conflicts may be configured dynamically anyway.

> This might seem restrictive, as you won't be able to define your own private application-specific data types with fixed identifiers, but this is necessary to statically guarantee collision-free coexistence of nodes from different vendors that use private data types. The core idea is that if you want a fixed ID, it must be public, so that other vendors can see it and avoid reusing it.

When nodes from different vendors use application-specific data types they will be configured dynamically. At the end of the day, the integrator will have control of what Port IDs they use, and can avoid using the same Port IDs allocated for other things (both dynamically and with the default port ID for custom messages).

On the other hand, checking that no definitions in the vendor specific (or standard) repo use these IDs are essential.

Reply:
> By rejecting definitions of default service/subject ID in the application specific domain will make the life of groups with full control of the nodes harder

This is true, but if we were to lift the restriction, such groups would have to remain committed to being fully isolated from the rest of the ecosystem once they started using private types with fixed ID, otherwise addition of any third-party (e.g. COTS) unit can potentially cause a conflict. Knowing that addition of foreign third-party hardware may cause a conflict will encourage divergence and fragmentation of the ecosystem, which we must avoid. We might sort of work around this by defining complex rules, such as "if a device is intended to be used outside of the vendor's ecosystem, it shall not use fixed identifiers in the non-regulated range", but I don't expect such a rule to work well in practice, if at all (because it would be difficult to enforce automatically, and humans are bad at following rules).

We might come up with some clever solutions to this later in UAVCAN v1.x. Remember, it is easier to lift restrictions than to introduce them later in a backward compatible way.

> At the end of the day, the integrator will have control of what Port IDs they use, and can avoid using the same Port IDs allocated for other things (both dynamically and with the default port ID for custom messages).

As per our discussion, a fixed port ID is a guarantee that the data type will always be accessible at this port ID. It is possible to use it with different ports as well, though. Context: https://forum.uavcan.org/t/data-type-compatibility-assurance/196/7?u=pavel.kirienko

Reply:
> This is true, but if we were to lift the restriction, such groups would have to remain committed to being fully isolated from the rest of the ecosystem once they started using private types with fixed ID, otherwise addition of any third-party (e.g. COTS) unit can potentially cause a conflict.

This is not true. Any third party unit would either use the vendor specific DTS or be configurable inside the application specific range. They would always be able to configure the new unit to IDs they're not using. If not, any COTS unit might also cause conflict with COTS units from a different brand.

If they were starting to sell the unit (instead of just the integrated application) they would have to move the definition to the vendor specific domain and request a new static ID (or use dynamically configured IDs). They would have to move their definition to the vendor specific namespace either way (to make it available and check out on Port compatibility checks) which would be impossible with an ID from the application specific DTS.

> We might sort of work around this by defining complex rules, such as "if a device is intended to be used outside of the vendor's ecosystem, it shall not use fixed identifiers in the non-regulated range", but I don't expect such a rule to work well in practice, if at all (because it would be difficult to enforce automatically, and humans are bad at following rules).

We only need one simple rule. The application integrator always decides the IDs inside the application DTS. If it's impossible to assign default IDs for types that only make sense to be used in one single application people are going to work around it by defining the ID as an associated constant, which results in the same problem + frustration.